runtime/src/accounts_db.rs

```diff
@@ -2466,12 +2466,19 @@ impl AccountsDb {
     }
 
     pub fn shrink_all_slots(&self, is_startup: bool) {
+        const MAX_DIRTY_STORES: usize = 10_000;
         if is_startup && self.caching_enabled {
             let slots = self.all_slots_in_storage();
-            let chunk_size = std::cmp::max(slots.len() / 8, 1); // approximately 400k slots in a snapshot
-            slots.par_chunks(chunk_size).for_each(|slots| {
-                for slot in slots {
-                    self.shrink_slot_forced(*slot, is_startup);
+            let outer_chunk_size = 2000;
+            let chunk_size = std::cmp::max(outer_chunk_size / 8, 1);
+            slots.chunks(outer_chunk_size).for_each(|chunk| {
+                chunk.par_chunks(chunk_size).for_each(|slots| {
+                    for slot in slots {
```

Comment:
So the idea of this structuring is that you want to work in parallel on slots within a group of 2000?
It looks like self.all_slots_in_storage() returns unordered slots from a dashmap.
Can you help me understand the thinking of why you break the slots into 2000 chunks? Maybe to do smaller cleans?

Reply:
@jeffwashington it's to consume and release the stores in `dirty_stores` so that minimize the number of open files

Reply:
So we limit the # of calls to shrink_slot_forced to 8 at a time.
And, we limit the total calls to shrink_slot_forced to 2k calls before we call clean_accounts.
So the idea is shrink causes us to create a new file and have the old one open at the same time while we create a new store? This was at most 8 in parallel before and still remains 8. At the end of shrink_slot_forced, presumably the old file is then closed and unmapped and recycled or deleted?
Is it important that the clean occurs once every ~2k shrinks and not on more slots at a time?
I'm trying to understand where the # of open files changes because of the 2k chunks.

Reply:
@jeffwashington the `dirty_stores` are like a set of "pending work" that is added to by the shrink, then consumed all  every call to `self.clean_accounts`, so the size of this pending work is now capped to roughly 2k items because we interleave calls to clean with the shrink, instead of shrinking all the slots at once and accumulating files in  `dirty_stores` without an intermediate clean.

Reply:
@carllin so why is DIRTY_STORES_CLEANING_THRESHOLD 10k? Doesn't that seem to mean we are only running clean when the amount of work is > 10k items?
And I am guessing the insert into dirty_stores that does a clone is why we are holding open mmapped files.
`self.dirty_stores.insert((slot, *store_id), store.clone());`

Reply:
> Doesn't that seem to mean we are only running clean when the amount of work is > 10k items

Yup we will only clean if it's bigger than that limit, but note this is only running at startup `shrink_all_slots`. I think we picked 10k as the interval because it's a lot smaller than the 700k we were seeing being accumulated on restart on MB 😃 

> And I am guessing the insert into dirty_stores that does a clone is why we are holding open mmapped files.

Yup exactly!